titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java

```diff
@@ -0,0 +1,262 @@
+                break;
+            case TaskStatus.REASON_STUCK_IN_STATE:
+            case TaskStatus.REASON_LOCAL_SYSTEM_ERROR:
+                errorKind = ErrorKind.LocalStartingError;
+                break;
+            case TaskStatus.REASON_TRANSIENT_SYSTEM_ERROR:
+            case TaskStatus.REASON_UNKNOWN_SYSTEM_ERROR:
+                errorKind = ErrorKind.SystemStartingError;
+                break;
+        }
+
+        PlacementHistory placement = tryGetOrCreatePlacementHistory(task);
+        if (placement != null) {
+            placement.onTaskFailure(errorKind);
+        }
+    }
+
+    private PlacementHistory tryGetOrCreatePlacementHistory(Task task) {
+        String hostId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_HOST);
+        if (hostId == null) {
+            invariants.inconsistent("Task without host name assigned: taskId=%s", task.getId());
+            return null;
+        }
+
+        PlacementHistory placement = agentsPlacementHistory.get(hostId);
+        if (placement == null) {
+            String instanceId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_INSTANCE_ID);
+            if (instanceId == null) {
+                invariants.inconsistent("Task without agent id assigned: taskId=%s", task.getId());
+                return null;
+            }
+            placement = agentManagementService.findAgentInstance(instanceId).map(PlacementHistory::new).orElse(null);
+        }
+
+        return placement;
+    }
+
+    private boolean isStartedTask(Task task) {
+        return JobFunctions.findTaskStatus(task, TaskState.Started).isPresent();
+    }
+
+    private enum ErrorKind {
+        LostAfterStarted,
+        LostBeforeStarted,
+        LocalStartingError,
+        SystemStartingError,
+        ContainerCrash
+    }
+
+    private class PlacementHistory {
+
+        private final AgentInstance instance;
+        private final TimeSeriesData errorTimeSeries = new TimeSeriesData(
+                configuration.getContainerFailureTrackingRetentionMs(),
+                1_000,
+                (value, delayMs) -> {
+                    long delayMin = 1 + delayMs / 60_000;
+                    return value / delayMin;
+                },
+                titusRuntime.getClock()
+        );
+
+        private PlacementHistory(AgentInstance instance) {
+            this.instance = instance;
+        }
+
+        private AgentInstance getInstance() {
+            return instance;
+        }
+
+        private double quality() {
+            double failureLevel = errorTimeSeries.getAggregatedValue();
+            return Math.max(0, Math.min(1, 1 - failureLevel));
+        }
+
+        private void onTaskStarted() {
+            errorTimeSeries.clear();
+        }
+
+        private void onTaskFailure(ErrorKind errorKind) {
+            double weight;
+            switch (errorKind) {
```

Comment:
I think you should move each weight to a constant and provide a javadoc mentioning the reasoning that went into choosing the weight. For instance, why is `LocalStartingError` a higher weight than `LostAfterStarted`?

Reply:
Done